waku/v2/protocol/waku_noise/noise.nim

```diff
@@ -98,10 +261,333 @@ proc genKeyPair*(rng: var BrHmacDrbgContext): KeyPair =
+  ss.ck = tempKeys[0]
+  ss.cs = CipherState(k: tempKeys[1])
+  trace "mixKey", ck = ss.ck, k = ss.cs.k
+
+# MixHash as per Noise specification http://www.noiseprotocol.org/noise.html#the-symmetricstate-object
+# Hashes data into a Symmetric State's handshake hash value h
+proc mixHash*(ss: var SymmetricState, data: openArray[byte]) =
+  # We prepare the hash context
+  var ctx: sha256
+  ctx.init()
+  # We add the previous handshake hash
+  ctx.update(ss.h.data)
+  # We append the input data
+  ctx.update(data)
+  # We hash and store the result in the Symmetric State's handshake hash value
+  ss.h = ctx.finish()
+  trace "mixHash", hash = ss.h.data
+
+# mixKeyAndHash as per Noise specification http://www.noiseprotocol.org/noise.html#the-symmetricstate-object
+# Combines MixKey and MixHash
+proc mixKeyAndHash*(ss: var SymmetricState, inputKeyMaterial: openArray[byte]) {.used.} =
+  var tempKeys: array[3, ChaChaPolyKey]
+  # Derives 3 keys using HKDF, the chaining key and the input key material
+  sha256.hkdf(ss.ck, inputKeyMaterial, [], tempKeys)
+  # Sets the chaining key
+  ss.ck = tempKeys[0]
+  # Updates the handshake hash value
+  ss.mixHash(tempKeys[1])
+  # Updates the Cipher state's key
+  # Note for later support of 512 bits hash functions: "If HASHLEN is 64, then truncates tempKeys[2] to 32 bytes."
+  ss.cs = CipherState(k: tempKeys[2])
+
+# EncryptAndHash as per Noise specification http://www.noiseprotocol.org/noise.html#the-symmetricstate-object
+# Combines encryptWithAd and mixHash
+proc encryptAndHash*(ss: var SymmetricState, plaintext: openArray[byte]): seq[byte]
+  {.raises: [Defect, NoiseNonceMaxError].} =
+  # The output ciphertext
+  var ciphertext: seq[byte]
+  # If an encryption key is set in the Symmetric state, we proceed with encryption using the handshake hash value as associated data
+  if ss.cs.hasKey:
+    ciphertext = ss.cs.encryptWithAd(ss.h.data, plaintext)
+  # According to specification, if no key is set return the plaintext
+  else:
+    ciphertext = @plaintext
+  # We call mixHash over the result
+  ss.mixHash(ciphertext)
+  return ciphertext
+
+# DecryptAndHash as per Noise specification http://www.noiseprotocol.org/noise.html#the-symmetricstate-object
+# Combines decryptWithAd and mixHash
+proc decryptAndHash*(ss: var SymmetricState, ciphertext: openArray[byte]): seq[byte]
+  {.raises: [Defect, NoiseDecryptTagError, NoiseNonceMaxError].} =
+  # The output plaintext
+  var plaintext: seq[byte]
+  # If an encryption key is set in the Symmetric state, we proceed with decryption using the handshake hash value as associated data
+  # Note that the ciphertext must contains an authorization tag, so its length should be greater or equal (in case of empty plaintext) the latter
+  if ss.cs.hasKey and ciphertext.len >= ChaChaPolyTag.len:
+    plaintext = ss.cs.decryptWithAd(ss.h.data, ciphertext)
+  # According to specification, if no key is set return the plaintext
+  else:
```

Comment:
Returning the input in case no key is set is a Noise specific condition and not of ChaChaPoly encryption/decryption. 
I just realized that indeed the check `ciphertext.len >= ChaChaPolyTag.len` is superflous here: `decryptWithAd` will read the last `ChaChaPolyTag.len` bytes as

`  let inputAuthorizationTag = ciphertext.toOpenArray(ciphertext.len - ChaChaPolyTag.len, ciphertext.high).intoChaChaPolyTag`

In turn, toOpenArray while computing the slice, will read starting from `max(ciphertext.len - ChaChaPolyTag.len, 0)` which is 0 in case `ciphertext.len >= ChaChaPolyTag.len`. Hence the full ciphertext is processed by `intoChaChaPolyTag`, which has such [check](https://github.com/status-im/nim-libp2p/blob/9973b9466dd062d7406d0178c750249b7e48471f/libp2p/crypto/chacha20poly1305.nim#L49) already implemented.

Reply:
This check is now moved in encryptWithAd/decryptWithAd. See https://github.com/status-im/nwaku/pull/933#discussion_r875832985